File: 06-linked_list/00287-Find_The_Duplicate_Number/00287-find_the_duplicate_number.go
// 287: Find the Duplicate Number
// https://leetcode.com/problems/find-the-duplicate-number/

package main

import "fmt"

// SOLUTION
func findDuplicate(nums []int) int {
    slow, fast := nums[0], nums[nums[0]]
        
    for slow != fast {
        slow = nums[slow]
        fast = nums[nums[fast]]
    }
    
    slow = 0
    for slow != fast {
        slow = nums[slow]
        fast = nums[fast]
    }

    return slow
}

func main() {
    // INPUT
    nums := []int{1,3,4,2,2}

    // OUTPUT
    result := findDuplicate(nums)
    fmt.Println(result)
}

File: 06-linked_list/00287-Find_The_Duplicate_Number/00287-find_the_duplicate_number.java
// 287: Find the Duplicate Number
// https://leetcode.com/problems/find-the-duplicate-number/


class Solution {
    // SOLUTION
    public int findDuplicate(int[] nums) {
        int slow = nums[0];
        int fast = nums[nums[0]];
        
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[nums[fast]];
        }
        
        slow = 0;
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }

        return slow;
    }


    public static void main(String[] args) {
        Solution o = new Solution();
        
        // INPUT
        int[] nums = {1,3,4,2,2};

        // OUTPUT
        var result = o.findDuplicate(nums);
        System.out.println(result);
    }
}